gallery_dl/extractor/tiktok.py

```diff
@@ -0,0 +1,98 @@
+INSENSITIVE_PHOTO = re.compile(re.escape("/photo/"), re.IGNORECASE)
+
+
+class TikTokExtractor(Extractor):
+    """Base class for TikTok extractors"""
+
+    category = "tiktok"
+    directory_fmt = ("{category}", "{user}")
+    filename_fmt = "{id}_{index}_{img_id}.{extension}"
+    archive_fmt = "{id}_{img_id}"
+    root = "https://www.tiktok.com/"
+    cookies_domain = ".tiktok.com"
+
+    def urls(self):
+        return [self.url]
+
+    def items(self):
+        for tiktok_url in self.urls():
+            # If we can recognise that this is a /photo/ link, preemptively
+            # replace it with /video/ to prevent a needless second request.
+            # See below.
+            tiktok_url = INSENSITIVE_PHOTO.sub("/video/", tiktok_url)
+            video_detail = util.json_loads(text.extr(
+                self.request(tiktok_url).text,
+                '<script id="__UNIVERSAL_DATA_FOR_REHYDRATION__" '
+                    'type="application/json">',
+                '</script>'
+            ))["__DEFAULT_SCOPE__"]
+            if "webapp.video-detail" not in video_detail:
+                # Only /video/ links result in the video-detail dict we need.
+                # Try again using that form of link.
+                tiktok_url = video_detail["seo.abtest"]["canonical"] \
+                    .replace("/photo/", "/video/")
+                video_detail = util.json_loads(text.extr(
+                    self.request(tiktok_url).text,
+                    '<script id="__UNIVERSAL_DATA_FOR_REHYDRATION__" '
+                        'type="application/json">',
+                    '</script>'
+                ))["__DEFAULT_SCOPE__"]
+            video_detail = video_detail["webapp.video-detail"]
+            if "statusMsg" in video_detail and \
+                video_detail["statusMsg"] == "author_secret":
+                raise exception.AuthorizationError("Login required to access "
+                                                   "this post")
+            post_info = video_detail["itemInfo"]["itemStruct"]
+            user = post_info["author"]["uniqueId"]
+            if "imagePost" in post_info:
+                yield Message.Directory, { "user": user }
+                img_list = post_info["imagePost"]["images"]
+                for i, img in enumerate(img_list):
+                    url = img["imageURL"]["urlList"][0]
+                    name_and_ext = text.nameext_from_url(url)
+                    yield Message.Url, url, {
+                        "id":        post_info["id"],
+                        "index":     i,
+                        "img_id":    name_and_ext["filename"].split("~")[0],
+                        "extension": name_and_ext["extension"],
+                        "width":     img["imageWidth"],
+                        "height":    img["imageHeight"]
+                    }
+            else:
+                # TODO: Not a slide show. Should pass this to yt-dlp.
```

Comment:
A couple questions regarding non-slideshow TikTok posts:

1. Should we support both photos and videos? My assumption is that we should, as extractors such as Instagram do.
2. If so, how should I go about supporting this? I know that I can `yield` a `Message.Queue` with a `ytdl:` URL, but this will end up downloading TikTok videos separately in a `ytdl/TikTok` category folder which does not seem ideal (as photos will be downloaded to `tiktok\{username}` and it makes the most sense to put videos in there).

Reply:
Also, I'd like to provide the option to *not* send video URLs to yt-dlp. I'm guessing I'd do this using the `--option` command-line argument?

Reply:
To have URLs get processed by ytdl, prepend the URL with `ytdl:`.
https://github.com/mikf/gallery-dl/blob/454f766f5ec249de93317610d375e84899f7b35c/gallery_dl/extractor/artstation.py#L66

Reply:
For a video option, use `self.config("videos")`
https://github.com/mikf/gallery-dl/blob/454f766f5ec249de93317610d375e84899f7b35c/gallery_dl/extractor/instagram.py#L58